### PR TITLE
Security: improve exact index matching performance

### PR DESCRIPTION
**Pull request by jaymode**
This commit improves the efficiency of exact index name matching by
separating exact matches from those that include wildcards or regular
expressions. Internally, exact matching is done using a HashSet instead
of adding the exact matches to the automata. For the wildcard and
regular expression matches, the underlying implementation has not
changed.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java [hunk 1]
```diff
@@ -64,6 +64,38 @@ public IndicesPermission(Group... groups) {
     }
 
     static Predicate<String> indexMatcher(List<String> indices) {
+        Set<String> exactMatch = new HashSet<>();
+        List<String> nonExactMatch = new ArrayList<>();
+        for (String indexPattern : indices) {
+            if (indexPattern.startsWith("/")) {
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java]: The existing `Automatons` code throws an exception if a pattern starts with `/` but doesn't end with `/`.
So we could simplify all this checking down to a single if/else and still be compatible with existing behaviour.
```
if (indexPattern.startsWith("/") || indexPattern.contains("*") || indexPattern.contains("?")) {
    nonExactMatch.add(indexPattern);
} else {
    exactMatch.add(indexPattern);
}
```

I don't care strongly, but shorter and simpler seems nicer.

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by tvernum** [comment]
I think as a follow up, when we have more time, we should:

1. Centralise this behaviour, and use it whereever we do pattern matching. It might mean putting a new `Patterns` class in front of `Automatons` that handles this logic.
2. Do some benchmarking about whether `"*xyz*"` is more efficient to run as `str.contains("xyz")` (and similarly `startsWith` and `endsWith`) and include that into the hypothetical `Patterns` utility class. I suspect, without testing that we pay a high price for using `Automaton` for trivial patterns.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java (hunk 1 above)
Comment (jaymode) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java]: ++ thanks for noticing that. I misread the if statement in Automatons

**Pull request merged**